SG panics when UnmarshalingJSON on RevTree if number of channels does not match number of revisions

Added a failing unit test that causes SG to Panic if the RevTree revs list is longer than the channels list.

https://github.com/couchbase/sync_gateway/blob/feature/issue_2737/db/revtree_test.go#L213

This is the associated panic

```
--- FAIL: TestRevTreeUnmarshalRevChannelCountMismatch (0.00s)
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range

goroutine 2077 [running]:
panic(0x567600, 0xc4200120f0)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
testing.tRunner.func1(0xc4217f2c00)
	/usr/local/go/src/testing/testing.go:579 +0x25d
panic(0x567600, 0xc4200120f0)
	/usr/local/go/src/runtime/panic.go:458 +0x243
encoding/json.(*decodeState).unmarshal.func1(0xc4203f1e88)
	/usr/local/go/src/encoding/json/decode.go:155 +0x11b
panic(0x567600, 0xc4200120f0)
	/usr/local/go/src/runtime/panic.go:458 +0x243
github.com/couchbase/sync_gateway/db.RevTree.UnmarshalJSON(0xc421808900, 0xc42102bf00, 0x79, 0x80, 0x80, 0xc42102bf00)
	/Users/andy/projects/sync_gateway/godeps/src/github.com/couchbase/sync_gateway/db/revtree.go:113 +0x536
github.com/couchbase/sync_gateway/db.(*RevTree).UnmarshalJSON(0xc421316560, 0xc42102bf00, 0x79, 0x80, 0xc420afce00, 0xbbf410)
	<autogenerated>:4 +0x70
encoding/json.(*decodeState).object(0xc4216aa000, 0x5edee0, 0xc421316560, 0x16)
	/usr/local/go/src/encoding/json/decode.go:562 +0x1cb5
encoding/json.(*decodeState).value(0xc4216aa000, 0x5edee0, 0xc421316560, 0x16)
	/usr/local/go/src/encoding/json/decode.go:368 +0x2cc
encoding/json.(*decodeState).unmarshal(0xc4216aa000, 0x5edee0, 0xc421316560, 0x0, 0x0)
	/usr/local/go/src/encoding/json/decode.go:169 +0x1fa
encoding/json.Unmarshal(0xc42102bf00, 0x79, 0x80, 0x5edee0, 0xc421316560, 0x80, 0x14)
	/usr/local/go/src/encoding/json/decode.go:97 +0x102
github.com/couchbase/sync_gateway/db.testUnmarshal(0xc4217f2c00, 0x625e82, 0x79, 0xc4213e1798)
	/Users/andy/projects/sync_gateway/godeps/src/github.com/couchbase/sync_gateway/db/revtree_test.go:173 +0xdc
github.com/couchbase/sync_gateway/db.TestRevTreeUnmarshalRevChannelCountMismatch(0xc4217f2c00)
	/Users/andy/projects/sync_gateway/godeps/src/github.com/couchbase/sync_gateway/db/revtree_test.go:215 +0x44
testing.tRunner(0xc4217f2c00, 0x6ad840)
	/usr/local/go/src/testing/testing.go:610 +0x81
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:646 +0x2ec

```

--- a/db/revtree.go
+++ b/db/revtree.go
@@ -100,6 +100,11 @@ func (tree RevTree) UnmarshalJSON(inputjson []byte) (err error) {
 		return
 	}
 
+	//validate revTreeList revs, parents and channels lists are of equal length
+	if !(len(rep.Revs) == len(rep.Parents) && len(rep.Revs) == len(rep.Channels)) {
+		return errors.New("revtreelist data is invalid, revs/parents/channels counts are inconsistent")
+	}
+
 	for i, revid := range rep.Revs {
 		info := RevInfo{ID: revid}
 		if rep.BodyMap != nil {

--- a/db/revtree_test.go
+++ b/db/revtree_test.go
@@ -20,6 +20,7 @@ import (
 	"github.com/couchbase/sync_gateway/base"
 	"github.com/couchbaselabs/go.assert"
 	"log"
+	"errors"
 )
 
 // 1-one -- 2-two -- 3-three
@@ -210,10 +211,11 @@ func TestRevTreeUnmarshal(t *testing.T) {
 	fmt.Printf("Unmarshaled to %v\n", gotmap)
 }
 
-func failTestRevTreeUnmarshalRevChannelCountMismatch(t *testing.T) {
+func TestRevTreeUnmarshalRevChannelCountMismatch(t *testing.T) {
 	const testJSON = `{"revs": ["3-three", "2-two", "1-one"], "parents": [1, 2, -1], "bodymap": {"0":"{}"}, "channels": [null, ["ABC", "CBS"]]}`
-	gotmap := testUnmarshal(t, testJSON)
-	fmt.Printf("Unmarshaled to %v\n", gotmap)
+	gotmap := RevTree{}
+	err := json.Unmarshal([]byte(testJSON), &gotmap)
+	assert.DeepEquals(t, err, errors.New("revtreelist data is invalid, revs/parents/channels counts are inconsistent"))
 }
 
 func TestRevTreeMarshal(t *testing.T) {

--- a/db/revtree_test.go
+++ b/db/revtree_test.go
@@ -210,6 +210,12 @@ func TestRevTreeUnmarshal(t *testing.T) {
 	fmt.Printf("Unmarshaled to %v\n", gotmap)
 }
 
+func failTestRevTreeUnmarshalRevChannelCountMismatch(t *testing.T) {
+	const testJSON = `{"revs": ["3-three", "2-two", "1-one"], "parents": [1, 2, -1], "bodymap": {"0":"{}"}, "channels": [null, ["ABC", "CBS"]]}`
+	gotmap := testUnmarshal(t, testJSON)
+	fmt.Printf("Unmarshaled to %v\n", gotmap)
+}
+
 func TestRevTreeMarshal(t *testing.T) {
 	bytes, err := json.Marshal(testmap)
 	assertNoError(t, err, "Couldn't write RevTree to JSON")

